Sources/ClientRuntime/EventStream/DefaultMessageEncoderStream.swift

```diff
@@ -0,0 +1,151 @@
+        let messageEncoder: MessageEncoder
+        let messageSinger: MessageSigner
+        let requestEncoder: RequestEncoder
+        var readAsyncIterator: AsyncIterator? = nil
+
+        public init(stream: AsyncThrowingStream<Event, Error>,
+                    messageEncoder: MessageEncoder,
+                    requestEncoder: RequestEncoder,
+                    messageSinger: MessageSigner) {
+            self.stream = stream
+            self.messageEncoder = messageEncoder
+            self.messageSinger = messageSinger
+            self.requestEncoder = requestEncoder
+            self.readAsyncIterator = makeAsyncIterator()
+        }
+
+        public struct AsyncIterator: AsyncIteratorProtocol {
+            let stream: AsyncThrowingStream<Event, Error>
+            let messageEncoder: MessageEncoder
+            var messageSinger: MessageSigner
+            let requestEncoder: RequestEncoder
+
+            private var lastMessageSent: Bool = false
+
+            init(stream: AsyncThrowingStream<Event, Error>,
+                 messageEncoder: MessageEncoder,
+                 requestEncoder: RequestEncoder,
+                 messageSinger: MessageSigner) {
+                self.stream = stream
+                self.messageEncoder = messageEncoder
+                self.messageSinger = messageSinger
+                self.requestEncoder = requestEncoder
+            }
+
+            mutating public func next() async throws -> Data? {
+                var iterator = stream.makeAsyncIterator()
+                guard let event = try await iterator.next() else {
+                    // There are no more messages in the base stream
+                    // if we have not sent the last message, send it now
+                    guard lastMessageSent else {
+                        let emptySignedMessage = try await messageSinger.signEmpty()
+                        let data = try messageEncoder.encode(message: emptySignedMessage)
+                        lastMessageSent = true
+                        return data
+                    }
+
+                    // mark the stream as complete
+                    return nil
+                }
+
+                // marshall event to message
+                let message = try event.marshall(encoder: requestEncoder)
+
+                // sign the message
+                let signedMessage = try await messageSinger.sign(message: message)
+
+                // encode again the signed message
+                let data = try messageEncoder.encode(message: signedMessage)
+                return data
+            }
+        }
+
+        public func makeAsyncIterator() -> AsyncIterator {
+            AsyncIterator(stream: stream,
+                          messageEncoder: messageEncoder,
+                          requestEncoder: requestEncoder,
+                          messageSinger: messageSinger)
+        }
+
+        // MARK: Stream
+
+        public var position: ClientRuntime.Data.Index = 0
```

Comment:
Can we add doc comments for all public properties and functions? I imagine we have corresponding docs on the protocol but we must bring also add comments here as well.

Reply:
fixed.